File: TrustoraAndroidApp/app/src/main/java/com/trustora/app/core/utils/TrustoraLocationCatalog.kt
package com.trustora.app.core.utils

import android.content.Context
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.trustora.app.core.models.DashboardLocationCity
import com.trustora.app.core.models.DashboardLocationCountry
import com.trustora.app.core.models.DashboardLocationState
import java.util.Locale

object TrustoraCompanyIdentificationTypes {
    // Mirrors the iOS/Next mapping used in company settings.
    val byCountryIso: Map<String, String> = mapOf(
        "RO" to "CUI", "GB" to "CRN", "DE" to "USt-IdNr", "FR" to "SIRET", "IT" to "P.IVA",
        "ES" to "NIF", "NL" to "RSIN", "PL" to "NIP", "BG" to "UIC", "HU" to "Adószám",
        "AT" to "UID", "BE" to "BCE / KBO", "DK" to "CVR", "SE" to "Org.nr", "NO" to "Org.nr",
        "FI" to "Y-tunnus", "PT" to "NIPC", "GR" to "AFM", "IE" to "CRO / VAT", "CZ" to "IČO",
        "SK" to "IČO", "SI" to "MŠ", "HR" to "OIB", "EE" to "Registrikood", "LV" to "Reģ. Nr.",
        "LT" to "Įm. k.", "CY" to "TIC", "MT" to "VAT / C", "LU" to "RCS", "CH" to "UID / CHE",
        "IS" to "Kennitala", "RS" to "PIB", "TR" to "VKN", "UA" to "EDRPOU", "US" to "EIN",
        "CA" to "BN", "MX" to "RFC", "BR" to "CNPJ", "AR" to "CUIT", "CL" to "RUT", "CO" to "NIT",
        "PE" to "RUC", "CN" to "USCI", "JP" to "CN", "IN" to "GSTIN / PAN", "KR" to "BRN",
        "SG" to "UEN", "AU" to "ABN", "NZ" to "NZBN", "HK" to "BRN", "TW" to "BAN", "ID" to "NPWP",
        "MY" to "SSM", "TH" to "TIN", "VN" to "MST", "AE" to "TRN", "SA" to "VAT / CR",
        "IL" to "H.P.", "ZA" to "CIPC", "EG" to "TRN", "NG" to "RC", "MA" to "ICE",
    )
}

class TrustoraLocationCatalog(context: Context) {
    private val appContext = context.applicationContext

    @Volatile
    private var loaded = false
    private var countries: List<DashboardLocationCountry> = emptyList()
    private var states: List<DashboardLocationState> = emptyList()
    private var cities: List<DashboardLocationCity> = emptyList()

    @Synchronized
    private fun ensureLoaded() {
        if (loaded) return

        countries = loadCountries()
        states = loadStates()
        cities = loadCities()
        loaded = true
    }

    fun allCountries(): List<DashboardLocationCountry> {
        ensureLoaded()
        return countries
    }

    fun statesOf(countryIso: String): List<DashboardLocationState> {
        ensureLoaded()
        val normalized = countryIso.trim().uppercase(Locale.ROOT)
        if (normalized.isEmpty()) return emptyList()
        return states
            .asSequence()
            .filter { it.countryCode == normalized }
            .sortedBy { it.name.lowercase(Locale.ROOT) }
            .toList()
    }

    fun citiesOf(countryIso: String, stateIso: String): List<DashboardLocationCity> {
        ensureLoaded()
        val normalizedCountry = countryIso.trim().uppercase(Locale.ROOT)
        val normalizedState = stateIso.trim().uppercase(Locale.ROOT)
        if (normalizedCountry.isEmpty() || normalizedState.isEmpty()) return emptyList()

        return cities
            .asSequence()
            .filter { it.countryCode == normalizedCountry && it.stateCode == normalizedState }
            .sortedBy { it.name.lowercase(Locale.ROOT) }
            .toList()
    }

    fun countryMatching(value: String): DashboardLocationCountry? {
        ensureLoaded()
        val normalized = value.trim()
        if (normalized.isEmpty()) return null

        countries.firstOrNull { it.isoCode.equals(normalized, ignoreCase = true) }?.let { return it }
        return countries.firstOrNull { it.name.equals(normalized, ignoreCase = true) }
    }

    fun normalizeStateIso(countryIso: String, stateValue: String?): String {
        val raw = stateValue?.trim().orEmpty()
        if (raw.isEmpty()) return ""

        val countryStates = statesOf(countryIso)
        countryStates.firstOrNull { it.isoCode.equals(raw, ignoreCase = true) }?.let { return it.isoCode }
        countryStates.firstOrNull { it.name.equals(raw, ignoreCase = true) }?.let { return it.isoCode }
        return raw
    }

    private fun loadCountries(): List<DashboardLocationCountry> {
        val rows = parseArrayOfObjects("location/country.json")
        return rows.mapNotNull { row ->
            val isoCode = row.string("isoCode")?.trim().orEmpty()
            val name = row.string("name")?.trim().orEmpty()
            if (isoCode.isEmpty() || name.isEmpty()) return@mapNotNull null
            DashboardLocationCountry(
                isoCode = isoCode.uppercase(Locale.ROOT),
                name = name,
                flag = row.string("flag").orEmpty(),
            )
        }.sortedBy { it.name.lowercase(Locale.ROOT) }
    }

    private fun loadStates(): List<DashboardLocationState> {
        val rows = parseArrayOfObjects("location/state.json")
        return rows.mapNotNull { row ->
            val isoCode = row.string("isoCode")?.trim().orEmpty()
            val name = row.string("name")?.trim().orEmpty()
            val countryCode = row.string("countryCode")?.trim().orEmpty()
            if (isoCode.isEmpty() || name.isEmpty() || countryCode.isEmpty()) return@mapNotNull null
            DashboardLocationState(
                isoCode = isoCode.uppercase(Locale.ROOT),
                name = name,
                countryCode = countryCode.uppercase(Locale.ROOT),
            )
        }
    }

    private fun loadCities(): List<DashboardLocationCity> {
        val json = readAsset("location/city.json")
        val root = JsonParser.parseString(json)
        if (!root.isJsonArray) return emptyList()

        val seen = hashSetOf<String>()
        val list = arrayListOf<DashboardLocationCity>()
        val array = root.asJsonArray
        for (index in 0 until array.size()) {
            val row = array[index]
            if (!row.isJsonArray) continue
            val values = row.asJsonArray
            if (values.size() < 3) continue

            val name = values[0].asStringOrNull()?.trim().orEmpty()
            val countryCode = values[1].asStringOrNull()?.trim()?.uppercase(Locale.ROOT).orEmpty()
            val stateCode = values[2].asStringOrNull()?.trim()?.uppercase(Locale.ROOT).orEmpty()
            if (name.isEmpty() || countryCode.isEmpty() || stateCode.isEmpty()) continue

            val unique = "$countryCode|$stateCode|${name.lowercase(Locale.ROOT)}"
            if (!seen.add(unique)) continue

            list += DashboardLocationCity(
                name = name,
                countryCode = countryCode,
                stateCode = stateCode,
            )
        }

        return list
    }

    private fun parseArrayOfObjects(assetPath: String): List<JsonObject> {
        val json = readAsset(assetPath)
        val root = JsonParser.parseString(json)
        if (!root.isJsonArray) return emptyList()
        val out = arrayListOf<JsonObject>()
        val array = root.asJsonArray
        for (index in 0 until array.size()) {
            val row = array[index]
            if (row.isJsonObject) out += row.asJsonObject
        }
        return out
    }

    private fun readAsset(path: String): String {
        return appContext.assets.open(path).bufferedReader().use { it.readText() }
    }

    private fun JsonObject.string(key: String): String? {
        val value = get(key) ?: return null
        return value.asStringOrNull()
    }
}

File: TrustoraApp/TrustoraApp/Core/Utilities/TrustoraLocationCatalog.swift
import Foundation

struct TrustoraLocationCountry: Identifiable, Hashable {
    let isoCode: String
    let name: String
    let flag: String

    var id: String { isoCode }
}

struct TrustoraLocationState: Identifiable, Hashable {
    let isoCode: String
    let name: String
    let countryCode: String

    var id: String { "\(countryCode)-\(isoCode)" }
}

struct TrustoraLocationCity: Identifiable, Hashable {
    let name: String
    let countryCode: String
    let stateCode: String

    var id: String { "\(countryCode)-\(stateCode)-\(name.lowercased())" }
}

enum TrustoraLocationCatalogError: LocalizedError {
    case missingResource(String)
    case invalidFormat(String)

    var errorDescription: String? {
        switch self {
        case let .missingResource(name):
            return "Missing location resource: \(name)"
        case let .invalidFormat(name):
            return "Invalid location resource format: \(name)"
        }
    }
}

enum TrustoraCompanyIdentificationTypes {
    // Matches Next.js COUNTRY_ID_TYPES mapping from company settings dialog.
    static let byCountryISO: [String: String] = [
        "RO": "CUI", "GB": "CRN", "DE": "USt-IdNr", "FR": "SIRET", "IT": "P.IVA",
        "ES": "NIF", "NL": "RSIN", "PL": "NIP", "BG": "UIC", "HU": "Adószám",
        "AT": "UID", "BE": "BCE / KBO", "DK": "CVR", "SE": "Org.nr", "NO": "Org.nr",
        "FI": "Y-tunnus", "PT": "NIPC", "GR": "AFM", "IE": "CRO / VAT", "CZ": "IČO",
        "SK": "IČO", "SI": "MŠ", "HR": "OIB", "EE": "Registrikood", "LV": "Reģ. Nr.",
        "LT": "Įm. k.", "CY": "TIC", "MT": "VAT / C", "LU": "RCS", "CH": "UID / CHE",
        "IS": "Kennitala", "RS": "PIB", "TR": "VKN", "UA": "EDRPOU", "US": "EIN",
        "CA": "BN", "MX": "RFC", "BR": "CNPJ", "AR": "CUIT", "CL": "RUT", "CO": "NIT",
        "PE": "RUC", "CN": "USCI", "JP": "CN", "IN": "GSTIN / PAN", "KR": "BRN",
        "SG": "UEN", "AU": "ABN", "NZ": "NZBN", "HK": "BRN", "TW": "BAN", "ID": "NPWP",
        "MY": "SSM", "TH": "TIN", "VN": "MST", "AE": "TRN", "SA": "VAT / CR",
        "IL": "H.P.", "ZA": "CIPC", "EG": "TRN", "NG": "RC", "MA": "ICE",
    ]
}

final class TrustoraLocationCatalog {
    static let shared = TrustoraLocationCatalog()

    private var hasLoaded = false
    private var countries: [TrustoraLocationCountry] = []
    private var states: [TrustoraLocationState] = []
    private var cities: [TrustoraLocationCity] = []

    private init() {}

    func allCountries() throws -> [TrustoraLocationCountry] {
        try ensureLoaded()
        return countries
    }

    func states(of countryISO: String) throws -> [TrustoraLocationState] {
        try ensureLoaded()
        let normalized = countryISO.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard !normalized.isEmpty else { return [] }

        return states
            .filter { $0.countryCode == normalized }
            .sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    func cities(of countryISO: String, stateISO: String) throws -> [TrustoraLocationCity] {
        try ensureLoaded()
        let normalizedCountry = countryISO.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        let normalizedState = stateISO.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard !normalizedCountry.isEmpty, !normalizedState.isEmpty else { return [] }

        return cities
            .filter { $0.countryCode == normalizedCountry && $0.stateCode == normalizedState }
            .sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    func country(matching value: String) throws -> TrustoraLocationCountry? {
        try ensureLoaded()
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else { return nil }

        if let byISO = countries.first(where: { $0.isoCode.caseInsensitiveCompare(normalized) == .orderedSame }) {
            return byISO
        }

        return countries.first(where: { $0.name.caseInsensitiveCompare(normalized) == .orderedSame })
    }

    func normalizeStateISO(countryISO: String, stateValue: String?) throws -> String {
        let raw = (stateValue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty else { return "" }

        let countryStates = try states(of: countryISO)
        if let byISO = countryStates.first(where: { $0.isoCode.caseInsensitiveCompare(raw) == .orderedSame }) {
            return byISO.isoCode
        }

        if let byName = countryStates.first(where: { $0.name.caseInsensitiveCompare(raw) == .orderedSame }) {
            return byName.isoCode
        }

        return raw
    }

    private func ensureLoaded() throws {
        guard !hasLoaded else { return }

        let loadedCountries = try loadCountries()
        let loadedStates = try loadStates()
        let loadedCities = try loadCities()

        countries = loadedCountries
        states = loadedStates
        cities = loadedCities
        hasLoaded = true
    }

    private func loadCountries() throws -> [TrustoraLocationCountry] {
        let data = try dataForResource(named: "country.json")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw TrustoraLocationCatalogError.invalidFormat("country.json")
        }

        return json.compactMap { item in
            guard let iso = (item["isoCode"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  let name = (item["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !iso.isEmpty,
                  !name.isEmpty
            else {
                return nil
            }

            let flag = (item["flag"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return TrustoraLocationCountry(isoCode: iso.uppercased(), name: name, flag: flag)
        }
        .sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    private func loadStates() throws -> [TrustoraLocationState] {
        let data = try dataForResource(named: "state.json")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw TrustoraLocationCatalogError.invalidFormat("state.json")
        }

        return json.compactMap { item in
            guard let iso = (item["isoCode"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  let name = (item["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  let country = (item["countryCode"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !iso.isEmpty,
                  !name.isEmpty,
                  !country.isEmpty
            else {
                return nil
            }

            return TrustoraLocationState(
                isoCode: iso.uppercased(),
                name: name,
                countryCode: country.uppercased()
            )
        }
    }

    private func loadCities() throws -> [TrustoraLocationCity] {
        let data = try dataForResource(named: "city.json")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [[Any]] else {
            throw TrustoraLocationCatalogError.invalidFormat("city.json")
        }

        var seen = Set<String>()
        var parsed: [TrustoraLocationCity] = []
        parsed.reserveCapacity(json.count)

        for item in json {
            guard item.count >= 3,
                  let rawName = item[0] as? String,
                  let rawCountryCode = item[1] as? String,
                  let rawStateCode = item[2] as? String
            else {
                continue
            }

            let name = rawName.trimmingCharacters(in: .whitespacesAndNewlines)
            let countryCode = rawCountryCode.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
            let stateCode = rawStateCode.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()

            guard !name.isEmpty, !countryCode.isEmpty, !stateCode.isEmpty else {
                continue
            }

            let uniqueKey = "\(countryCode)|\(stateCode)|\(name.lowercased())"
            guard seen.insert(uniqueKey).inserted else {
                continue
            }

            parsed.append(
                TrustoraLocationCity(
                    name: name,
                    countryCode: countryCode,
                    stateCode: stateCode
                )
            )
        }

        return parsed
    }

    private func dataForResource(named name: String) throws -> Data {
        let baseName = (name as NSString).deletingPathExtension
        let fileExtension = (name as NSString).pathExtension

        if let url = resourceURL(baseName: baseName, extension: fileExtension) {
            return try Data(contentsOf: url)
        }

        throw TrustoraLocationCatalogError.missingResource(name)
    }

    private func resourceURL(baseName: String, extension fileExtension: String) -> URL? {
        if let direct = Bundle.main.url(forResource: baseName, withExtension: fileExtension) {
            return direct
        }

        let candidateSubdirectories: [String?] = [
            "Resources/LocationData",
            "LocationData",
            "Resources",
            nil,
        ]

        for subdirectory in candidateSubdirectories {
            if let url = Bundle.main.url(
                forResource: baseName,
                withExtension: fileExtension,
                subdirectory: subdirectory
            ) {
                return url
            }
        }

        if let all = Bundle.main.urls(forResourcesWithExtension: fileExtension, subdirectory: nil) {
            return all.first { $0.lastPathComponent == "\(baseName).\(fileExtension)" }
        }

        return nil
    }
}
